runtime/hsa-runtime/core/driver/xdna/amd_xdna_driver.cpp

```diff
@@ -330,5 +347,255 @@ hsa_status_t XdnaDriver::FreeDeviceHeap() {
+    amdxdna_cmd *cmd = nullptr;
+    uint32_t cmd_size = sizeof(amdxdna_cmd) + pkt->count * sizeof(uint32_t);
+    if (CreateCmd(cmd_size, &cmd_bo_handle, &cmd))
+      return HSA_STATUS_ERROR;
+
+    // Filling in the fields of the command
+    cmd->state = pkt->state;
+    cmd->extra_cu_masks = 0;
+
+    // The driver places a structure before each command in a command chain.
+    // Need to increase the size of the command by the size of this structure.
+    cmd->count = pkt->count + CMD_COUNT_SIZE_INCREASE;
+    cmd->opcode = pkt->opcode;
+    cmd->data[0] = cmd_pkt_payload->cu_mask;
+    memcpy((cmd->data + 1), cmd_pkt_payload->data, 4 * pkt->count);
+
+    // Keeping track of the handle
+    cmd_handles.push_back(cmd_bo_handle);
+    cmds.push_back(cmd);
+    cmd_sizes.push_back(cmd_size);
+  }
+
+  // Creating a packet that contains the command chain
+  uint32_t cmd_chain_bo_handle = 0;
+  amdxdna_cmd *cmd_chain = nullptr;
+  int cmd_chain_size = (cmd_handles.size() + 1) * sizeof(uint32_t);
+  if (CreateCmd(cmd_chain_size, &cmd_chain_bo_handle, &cmd_chain))
+    return HSA_STATUS_ERROR;
+
+  // Writing information to the command buffer
+  amdxdna_cmd_chain *cmd_chain_payload =
+      reinterpret_cast<amdxdna_cmd_chain *>(cmd_chain->data);
+
+  // Creating a command chain
+  cmd_chain->state = HSA_AMD_AIE_ERT_STATE_NEW;
+  cmd_chain->extra_cu_masks = 0;
+  cmd_chain->count =
+      sizeof(amdxdna_cmd_chain) + cmd_handles.size() * sizeof(uint64_t);
+  cmd_chain->opcode = HSA_AMD_AIE_ERT_CMD_CHAIN;
+  cmd_chain_payload->command_count = cmd_handles.size();
+  cmd_chain_payload->submit_index = 0;
+  cmd_chain_payload->error_index = 0;
+  for (int i = 0; i < cmd_handles.size(); i++) {
+    cmd_chain_payload->data[i] = cmd_handles[i];
+  }
+
+  // Syncing BOs before we execute the command
+  if (SyncBos(bo_addrs, bo_sizes))
+    return HSA_STATUS_ERROR;
+
+  // Removing duplicates in the bo container. The driver will report
+  // an error if we provide the same BO handle multiple times.
+  // This can happen if any of the BOs are the same across jobs
+  std::sort(bo_args.begin(), bo_args.end());
+  bo_args.erase(std::unique(bo_args.begin(), bo_args.end()), bo_args.end());
+
+  // Filling in the fields to execute the command chain
+  amdxdna_drm_exec_cmd exec_cmd_0 = {};
+  exec_cmd_0.ext = 0;
+  exec_cmd_0.ext_flags = 0;
+  exec_cmd_0.hwctx = hw_ctx_handle;
+  exec_cmd_0.type = AMDXDNA_CMD_SUBMIT_EXEC_BUF;
+  exec_cmd_0.cmd_handles = cmd_chain_bo_handle;
+  exec_cmd_0.args = (uint64_t)bo_args.data();
+  exec_cmd_0.cmd_count = 1;
+  exec_cmd_0.arg_count = bo_args.size();
+
+  // Executing all commands in the command chain
+  ExecCmdAndWait(&exec_cmd_0, hw_ctx_handle);
+
+  // Unmapping and closing the cmd BOs
+  drm_gem_close close_bo_args{0};
+  for (int i = 0; i < cmd_handles.size(); i++) {
+    munmap(cmds[i], cmd_sizes[i]);
+    close_bo_args.handle = cmd_handles[i];
+    ioctl(fd_, DRM_IOCTL_GEM_CLOSE, &close_bo_args);
+  }
+
```

Comment:
It needs a check for munmap errors. 

```suggestion
  for (uint32_t i = 0; i < cmd_handles.size(); i++) {
    if (munmap(cmds[i], cmd_sizes[i]) != 0)
        return HSA_STATUS_ERROR;
    }
    
    drm_gem_close close_bo_args = {};
    close_bo_args.handle = cmd_handles[i];
    ioctl(fd_, DRM_IOCTL_GEM_CLOSE, &close_bo_args);
  }
```

Reply:
Addressed in e6f15f4.